src/Microsoft.OData.Core/ODataCollectionWriterCore.cs

```diff
@@ -508,20 +511,52 @@ private void StartPayloadInStartState()
             Scope current = this.scopes.Peek();
             if (current.State == CollectionWriterState.Start)
             {
-                this.InterceptException(this.StartPayload);
+                this.InterceptException((thisParam) => thisParam.StartPayload());
             }
         }
 
         /// <summary>
         /// Catch any exception thrown by the action passed in; in the exception case move the writer into
-        /// state ExceptionThrown and then rethrow the exception.
+        /// state Error and then re-throw the exception.
         /// </summary>
         /// <param name="action">The action to execute.</param>
-        private void InterceptException(Action action)
+        /// <remarks>
+        /// Make sure to only use anonymous functions that don't capture state from the enclosing context, 
+        /// so the compiler optimizes the code to avoid creating allocations on every call to this method.
+        /// </remarks>
+        private void InterceptException(Action<ODataCollectionWriterCore> action)
+        {
+            try
+            {
+                action(this);
+            }
+            catch
+            {
+                if (!IsErrorState(this.State))
+                {
+                    this.EnterScope(CollectionWriterState.Error, this.scopes.Peek().Item);
+                }
+
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state Error and then rethrow the exception.
+        /// </summary>
+        /// <typeparam name="TArg0">The action argument type.</typeparam>
+        /// <param name="action">The action to execute.</param>
+        /// <param name="arg0">The argument value provided to the action.</param>
+        /// <remarks>
+        /// Make sure to only use anonymous functions that don't capture state from the enclosing context, 
+        /// so the compiler optimizes the code to avoid creating allocations on every call to this method.
+        /// </remarks>
+        private void InterceptException<TArg0>(Action<ODataCollectionWriterCore, TArg0> action, TArg0 arg0)
         {
             try
             {
-                action();
+                action(this, arg0);

```

Comment:
Might be a good idea to indicate that this pattern is used as an optimization? Or?